kernel/src/checkpoint/log_replay.rs

```diff
@@ -0,0 +1,621 @@
+        let mut seen_txns = HashSet::new();
+        seen_txns.insert("app1".to_string());
+
+        let mut visitor = CheckpointVisitor::new(
+            &mut seen_file_keys,
+            true,
+            vec![true; 3],
+            0,
+            true,           // The visior has already seen a protocol action
+            true,           // The visitor has already seen a metadata action
+            &mut seen_txns, // Pre-populated transaction
+        );
+
+        visitor.visit_rows_of(batch.as_ref())?;
+
+        // All actions should be skipped as they have already been seen
+        let expected = vec![false, false, false];
+        assert_eq!(visitor.selection_vector, expected);
+        assert_eq!(visitor.total_non_file_actions, 0);
+        assert_eq!(visitor.file_actions_count, 0);
+
+        Ok(())
+    }
+
+    #[test]
+    fn test_checkpoint_visitor_duplicate_non_file_actions() -> DeltaResult<()> {
+        let json_strings: StringArray = vec![
+            r#"{"txn":{"appId":"app1","version":1,"lastUpdated":123456789}}"#,
+            r#"{"txn":{"appId":"app1","version":1,"lastUpdated":123456789}}"#, // Duplicate txn
+            r#"{"txn":{"appId":"app2","version":1,"lastUpdated":123456789}}"#, // Different app ID
+            r#"{"protocol":{"minReaderVersion":3,"minWriterVersion":7}}"#,
+            r#"{"protocol":{"minReaderVersion":3,"minWriterVersion":7}}"#, // Duplicate protocol
+            r#"{"metaData":{"id":"testId","format":{"provider":"parquet","options":{}},"schemaString":"{\"type\":\"struct\",\"fields\":[{\"name\":\"value\",\"type\":\"integer\",\"nullable\":true,\"metadata\":{}}]}","partitionColumns":[],"configuration":{},"createdTime":1677811175819}}"#,
+            // Duplicate metadata
+            r#"{"metaData":{"id":"testId","format":{"provider":"parquet","options":{}},"schemaString":"{\"type\":\"struct\",\"fields\":[{\"name\":\"value\",\"type\":\"integer\",\"nullable\":true,\"metadata\":{}}]}","partitionColumns":[],"configuration":{},"createdTime":1677811175819}}"#, 
+        ]
+        .into();
+        let batch = parse_json_batch(json_strings);
+
+        let mut seen_file_keys = HashSet::new();
+        let mut seen_txns = HashSet::new();
+        let mut visitor = CheckpointVisitor::new(
+            &mut seen_file_keys,
+            true, // is_log_batch
+            vec![true; 7],
+            0, // minimum_file_retention_timestamp
+            false,
+            false,
+            &mut seen_txns,
+        );
+
+        visitor.visit_rows_of(batch.as_ref())?;
+
+        // First occurrence of each type should be included
+        let expected = vec![true, false, true, true, false, true, false];
+        assert_eq!(visitor.selection_vector, expected);
+        assert_eq!(visitor.seen_txns.len(), 2); // Two different app IDs
+        assert_eq!(visitor.total_non_file_actions, 4); // 2 txns + 1 protocol + 1 metadata
+        assert_eq!(visitor.file_actions_count, 0);
+
+        Ok(())
```

Comment:
for many of these tests should we confirm that the original actions/data is the same (maybe at least len is same?)